### PR TITLE
fix(npm): set Yarn cache/global folder

### LINKED ISSUE #8343: Yarn: set YARN_CACHE_FOLDER
**What would you like Renovate to be able to do?**

Set `YARN_CACHE_FOLDER` to our usual cache location if:
- No `cacheFolder` specified in Yarn config
- `.yarn/cache` does not already exist

**Did you already have any implementation ideas?**

See https://github.com/renovatebot/renovate/issues/8247

### PR DESCRIPTION
**Pull request by ylemkimon**
## Changes:

- Sets [`YARN_CACHE_FOLDER`](https://classic.yarnpkg.com/en/docs/cli/cache#toc-change-the-cache-path-for-yarn) for Yarn 1
- Sets [`YARN_GLOBAL_FOLDER`](https://yarnpkg.com/configuration/yarnrc#globalFolder) for Yarn >=2

## Context:

Fixes #8343.

To use the global cache in Yarn 2, `YARN_ENABLE_GLOBAL_CACHE` should be set. However, it should not be set if zero-install is used (#11012), so it's not implemented in this PR.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository (https://github.com/ylemkimon/renovate-yarn-3 and https://github.com/ylemkimon/renovate-yarn)

### CODE REVIEW

**Review by rarkins — DISMISSED** [review]

File: lib/manager/npm/post-update/index.ts [hunk 1]
```diff
@@ -441,6 +441,10 @@ export async function getAdditionalFiles(
       './others/yarn',
       'YARN_CACHE_FOLDER'
     ),
+    YARN_GLOBAL_FOLDER: await ensureCacheDir(
+      './others/berry',
+      'YARN_GLOBAL_FOLDER'
+    ),
```
Comment (viceice) [inline lib/manager/npm/post-update/index.ts:447]: @rarkins This need to be adapted in v26

File: lib/manager/npm/post-update/yarn.spec.ts [hunk 2]
```diff
@@ -59,7 +59,14 @@ describe(getName(), () => {
         },
         postUpdateOptions: ['yarnDedupeFewer', 'yarnDedupeHighest'],
       };
-      const res = await yarnHelper.generateLockFile('some-dir', {}, config);
+      const res = await yarnHelper.generateLockFile(
+        'some-dir',
+        {
+          YARN_CACHE_FOLDER: 'yarn-cache-folder',
+          YARN_GLOBAL_FOLDER: 'yarn-global-folder',
```
Comment (viceice) [inline lib/manager/npm/post-update/yarn.spec.ts:66]: Can you use some tmp folder here, eg `/tmp/renovate/cache/yarn-global`

File: lib/manager/npm/post-update/index.ts (hunk 1 above)
Comment (ylemkimon) [reply lib/manager/npm/post-update/index.ts:447]: AFAIS this wouldn't be a breaking change.

Comment (viceice) [reply lib/manager/npm/post-update/index.ts:447]: Was just a note for our v26 branch. 😅

File: lib/manager/npm/post-update/yarn.spec.ts (hunk 2 above)
Comment (ylemkimon) [reply lib/manager/npm/post-update/yarn.spec.ts]: Done :smile:

File: lib/manager/npm/post-update/index.ts [hunk 3]
```diff
@@ -441,6 +441,10 @@ export async function getAdditionalFiles(
       './others/yarn',
       'YARN_CACHE_FOLDER'
     ),
+    YARN_GLOBAL_FOLDER: await ensureCacheDir(
+      './others/berry',
+      'YARN_GLOBAL_FOLDER'
```
Comment (viceice) [inline lib/manager/npm/post-update/index.ts:446]: @rarkins should we ignore predefined env, we will do it in v26 anyways. 🤔

Comment (rarkins) [reply lib/manager/npm/post-update/index.ts:446]: I don't mind, keep it as is if that's simplest

File: lib/manager/npm/post-update/index.ts (hunk 1 above)
Comment (viceice) [inline lib/manager/npm/post-update/index.ts:447]:
```suggestion
    YARN_GLOBAL_FOLDER: await ensureCacheDir('./others/berry'),
```

I would prefer this, so we don't raise expectations which will break in v26.

**Review by viceice — APPROVED** [review]

**Comment by ylemkimon** [comment]
@rarkins The CI didn't run due to issues with GitHub Actions yesterday (https://www.githubstatus.com/incidents/lc8s9rtyjmgq). So I've closed and reopened the PR and the auto-merge is disabled.

**Pull request merged**